coderd/provisionerdserver/acquirer.go

```diff
@@ -0,0 +1,481 @@
+	go func() {
+		defer close(subscribed)
+		eb := backoff.NewExponentialBackOff()
+		eb.MaxElapsedTime = 0 // retry indefinitely
+		eb.MaxInterval = dbMaxBackoff
+		bkoff := backoff.WithContext(eb, a.ctx)
+		var cancel context.CancelFunc
+		err := backoff.Retry(func() error {
+			cancelFn, err := a.ps.SubscribeWithErr(EventJobPosted, a.jobPosted)
+			if err != nil {
+				a.logger.Warn(a.ctx, "failed to subscribe to job postings", slog.Error(err))
+				return err
+			}
+			cancel = cancelFn
+			return nil
+		}, bkoff)
+		if err != nil {
+			if a.ctx.Err() == nil {
+				a.logger.Error(a.ctx, "code bug: retry failed before context canceled", slog.Error(err))
+			}
+			return
+		}
+		defer cancel()
+		bkoff.Reset()
+		a.logger.Debug(a.ctx, "subscribed to job postings")
+
+		// unblock the outer function from returning
+		subscribed <- struct{}{}
+
+		// hold subscriptions open until context is canceled
+		<-a.ctx.Done()
+	}()
+	<-subscribed
+}
+
+func (a *Acquirer) jobPosted(ctx context.Context, message []byte, err error) {
+	if xerrors.Is(err, pubsub.ErrDroppedMessages) {
+		a.logger.Warn(a.ctx, "pubsub may have dropped job postings")
+		a.clearOrPendAll()
+		return
+	}
+	if err != nil {
+		a.logger.Warn(a.ctx, "unhandled pubsub error", slog.Error(err))
+		return
+	}
+	posting := JobPosting{}
+	err = json.Unmarshal(message, &posting)
+	if err != nil {
+		a.logger.Error(a.ctx, "unable to parse job posting",
+			slog.F("message", string(message)),
+			slog.Error(err),
+		)
+		return
+	}
+	a.logger.Debug(ctx, "got job posting", slog.F("posting", posting))
+
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	for _, d := range a.q {
+		if d.contains(posting) {
+			a.clearOrPendLocked(d)
+			// we only need to wake up a single domain since there is only one
+			// new job available
+			return
+		}
+	}
+}
+
+func (a *Acquirer) clearOrPendAll() {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	for _, d := range a.q {
+		a.clearOrPendLocked(d)
+	}
+}
+
+func (a *Acquirer) clearOrPend(d domain) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	if len(d.acquirees) == 0 {
+		// this can happen if the domain is removed right around the time the
+		// backup poll (which calls this function) triggers.  Nothing to do
+		// since there are no acquirees.
+		return
+	}
+	a.clearOrPendLocked(d)
+}
+
+func (*Acquirer) clearOrPendLocked(d domain) {
+	// MUST BE CALLED HOLDING THE a.mu LOCK
+	var nominee *acquiree
+	for _, w := range d.acquirees {
+		if nominee == nil {
+			nominee = w
+		}
+		// acquiree in progress always takes precedence, since we don't want to
+		// wake up more than one acquiree per dKey at a time.
+		if w.inProgress {
+			nominee = w
+			break
+		}
+	}
+	if nominee.inProgress {
+		nominee.pending = true
+		return
+	}
+	nominee.inProgress = true
+	nominee.clearance <- struct{}{}
+}
+
+type dKey string
+
+func domainKey(pt []database.ProvisionerType, tags Tags) dKey {
+	sb := strings.Builder{}
+	for _, t := range pt {
+		_, _ = sb.WriteString(string(t))
+		_ = sb.WriteByte(0x00)
```

Comment:
Why a null byte as separator?

EDIT: after reading the change in `provisionerdserver.go`, this is more clear to me. Might be worth dropping a comment here though.